### DISCUSSION TITLE
v0.2.1 - AIM New Release

### DISCUSSION THREAD

**Discussion by Micro-Novelty** [Micro-Novelty/IntegratedPipeline-Specialized-Non-LLM-AI-Agent-Framework#14] [General]
## 🎯 Overview
IntegratedPipeline is a standalone Custom AI Agent Library for memory-augmented agentic frameworks, designed for continuous learning with MANN-Type Architecture designed for long term operation.

## 🎯 The provided Files below includes:
- Binaries for AbstractIntegratedmodule (In a zipped package file)
- Docker Quick start usage
- python script (Complete with basic instruction and basic prediction and asynchronous prediction setup) that was used by Docker quick start file.
- .txt file with CSV format, contain labels and titles, used for training and prediction (You can customize the labels and titles).
   Note: - for the .txt file, you must put it in the same directory as your/main.py python script (such as my main.py i provided)

## NEW SUPPORTING ARCHITECTURE!
[=] LSTM Architecture:
- LSTMs (Long Short-Term Memory) and Transformers are foundational deep learning architectures built to process sequential data (like text or time series). While both handle the flow of time and context, they do so using completely different mechanisms.
-  LSTM doesn't act as a Main orchestrator, instead it Provides coherent Short-term memory for the Ensemble architecture, acting as a support mechanism to provides proof-of-credibility of a given answer from past previous context input, this allows flexible and achievable Aggreement between Transformer and MLP.

### Core Features
- **Memory-Augmented Neural Networks (MANN)** - External dynamic memory module for persistent learning
- **Abstract Weight Encoder (AWE)** - Eigenvalue-based weight shaping for noise robustness
- **Specialized MLP** - Noise-robust classification with geometric weight adaptation
- **Optimized Transformer** - Alpha-based computing for stable contextual understanding
- **LSTM Architecture** - Efficient LSTM Using AWE Weight __init__.
- **Hybrid Ensemble** - Dynamic weighting of MLP + Transformer predictions
- **SQLite Integration** - Local database for memory, attention weights, and predictions
- **Peer-to-Peer Coordination** - Distributed agent communication with SSL security, Cohesive Interaction, and Compact multi-modal ensemble weighting from peer.

### Platform Support
- ✅ Windows (x86_64)
- ✅ Linux (x86_64)
- ✅ Raspberry Pi (ARM64 v7/v8) 

### Python Support
- Python 3.13 Only for windows 
- Python 3.12 for x86-64 machine
- Python 3.10 for ARM64 machine (Stable Robust version, Officially tested in ARM64 Environment)

##  📦 Library requirements when using binaries
- numpy
- scikit-learn
- pandas
- aiohttp
- psutil

## 📦 Binaries Included

Download the appropriate binary for your platform:
- `AbstractIntegratedModule.pyd` - Windows
- `AbstractIntegratedModule.cpython-39-x86_64-linux-gnu.so` - Linux x86_64
- `AbstractIntegratedModule.cpython-39-aarch64-linux-gnu.so` - Raspberry Pi ARM64

Download the appropriate files too:
- `Dockerfile + required files (see in Docker Application block in this repo)` (For Container assemble), can be downloaded in the code section 
- `main.py` (Script to execute in Quick_Docker_start)
- `Quick_Docker_start.sh` (For quick start)

## 📦 New Architecture
A. Main New components (Updated): 
- PipelineAsyncManager (For asynchronous prediction handling, fully local using worker based features)
- CohesiveAgentDeployment (for asynchronous prediction request from peer, Hybrid based feature, supports synchronous prediction + asynchronous prediction from peer)
- ConsecutivePeerAgent (P2P based architecture, is used for ensemble weighting from peers asynchronously)
B. Other Supporting Architectures
- ThreadedMessageQueue (For thread handling during asynchronous request)
- AsyncMessageQueue (For Asynchronous message  request handling, using worker based Queue features)

## Bug fixed
- fixed Undefined variables in CohesiveAgentDeployment rarely used function.
- P2P Flow in ARM64 Environment is verified and tested multiple Times, works efficiently.

## 🚀 Quick Start

```python
from AbstractIntegratedModule import IntegratedPipeline
from AbstractIntegratedModule import PipelinePredictionManager
import numpy as np

memory_name = 'agent_memory'
cert_file = <your_cert_file_dir> # your .crt file
key_file = <your_key_file_dir> # your .key file

main_model = IntegratedPipeline(memory_name=memory_name, use_async=True, agent_port=5000, ssl_cert_file=cert_file, ssl_key_file=key_file)
 # provide cert_file path or key_file path (optional)
main_prediction = PipelinePredictionManager(main_model, label_csv='example_manual_training.txt', target_title='window_title', label='label')
# example_manual_training is a .txt file that contain csv format like above example.

example_rules = [
                     # === WORK / PRODUCTIVITY ===
                     (r'code|programming|develop|debug|compile|script', 'focused_work'),
                     (r'vscode|visual_studio|ide|terminal|shell', 'focused_work'),
                     (r'notion|evernote|onenote|notes|todo|task', 'productive'),
                     (r'slack|teams|discord|zoom|meeting|call', 'communication'),
                     (r'email|gmail|outlook|inbox|mail', 'communication'),
                     
                     # === ENTERTAINMENT ===
                     (r'youtube|netflix|twitch|stream|video', 'entertainment'),
                     (r'music|spotify|soundcloud|audio|player', 'entertainment'),
                     (r'game|gaming|steam|epic|play', 'gaming'),
                     (r'facebook|instagram|tiktok|social|post', 'social_media'),
                     
                     # === BROWSING ===
                     (r'chrome|firefox|edge|safari|browser', 'browsing'),
                     (r'google|search|wiki|wiki|article', 'information'),
                     (r'stackoverflow|github|docs|documentation', 'research'),

                     # more rules
                 ]
# activate explainability capability to explain uncertainty:
main_model.show_explainability_details = True
main_model.distribution.predict_manager = main_prediction # set PipelinePredictionManager to AgentDistributedInference for asynchronous prediction later

# test samples with more sophisticated rules and more complex titles for prediction
# (title, intent)
test_titles = [
 ("Opening Thesis.docx", "slight_work"),
 ("Watching YouTube and Google Chrome", "distracted"),
 ("Watching Slack", "communication"),
 ("Programming in Visual Studio Code", "focused_work"),
 ("Watching netflix.com - Chrome", "break"),
# more titles 
 ]  
            
titles, y, label_map = main_prediction.load_labels_from_csv(<your_filename>, <target_title>, <target_label>)
# small training with simple titles
main_model.train(titles, y)
   
results, chosen_label, confidence = main_prediction.advanced_prediction_method(test_titles, label_map, example_rules,
                             show_proba=False, top_k=3, 
                             use_transformer=True,
                             return_attention=False,
                             save_results=True)

# ... more features/wrapper you can add
```
)

<hr /><em>This discussion was created from the release <a href='https://github.com/Micro-Novelty/IntegratedPipeline-Specialized-Non-LLM-AI-Agent-Framework/releases/tag/0.2.1-0.3.3'>v0.2.1 - AIM New Release</a>.</em>
